client_integration_test.go

```diff
@@ -0,0 +1,184 @@
+	}
+	if resp.URI == "" {
+		t.Errorf("Response URI should not be empty")
+	}
+	assertEqual(t, ref, *resp.Reference)
+	assertEqual(t, "Hello John,\n\nToday is Friday.", resp.Content.Body)
+}
+
+func TestSendEmail(t *testing.T) {
+	client, _, emailTemplateId := setup(t)
+
+	emailAddress := os.Getenv("EMAIL_RECIPIENT")
+	if emailAddress == "" {
+		t.Fatal("EMAIL_RECIPIENT environment variable not set")
+	}
+
+	ref := "TestSendEmail"
+	resp, err := client.SendEmail(
+		emailTemplateId,
+		emailAddress,
+		notify.Personalisation{
+			{"name", "John"},
+			{"colour", "pink"},
+		},
+		notify.Reference(ref),
+	)
+	if err != nil {
+		t.Fatalf("Error sending email: %v", err)
+	}
+
+	if resp.ID == "" {
+		t.Errorf("Response ID should not be empty")
+	}
+	if resp.URI == "" {
+		t.Errorf("Response URI should not be empty")
+	}
+	assertEqual(t, "Hi John,\n\nMy favourite colour is pink.\n", resp.Content.Body)
+}
+
+func TestGetAllTemplates(t *testing.T) {
+	client, smsTemplateId, emailTemplateId := setup(t)
+	resp, err := client.GetAllTemplates("")
+	if err != nil {
+		t.Fatalf("Error fetching all templates: %v", err)
+	}
+
+	assertTemplateFound(t, resp, []string{emailTemplateId, smsTemplateId})
+}
+
+func TestGetSMSTemplates(t *testing.T) {
+	client, smsTemplateId, _ := setup(t)
+	resp, err := client.GetAllTemplates("sms")
+	if err != nil {
+		t.Fatalf("Error fetching sms templates: %v", err)
+	}
+
+	assertTemplateFound(t, resp, []string{smsTemplateId})
+}
+
+func TestGetEmailTemplates(t *testing.T) {
+	client, _, emailTemplateId := setup(t)
+	resp, err := client.GetAllTemplates("email")
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
+	}
+
+	assertTemplateFound(t, resp, []string{emailTemplateId})
+}
+
+func TestGetTemplateById(t *testing.T) {
+	client, smsTemplateId, _ := setup(t)
+	resp, err := client.GetTemplateByID(smsTemplateId)
+	if err != nil {
+		t.Fatalf("Error fetching template by id: %v", err)
+	}
+	assertEqual(t, "go-sdk-test-sms", resp.Name)
+	assertEqual(t, "sms", resp.Type)
+}
+
+func TestGetTemplateByIdAndVersion(t *testing.T) {
+	client, _, emailTemplateId := setup(t)
+	resp, err := client.GetTemplateByIDAndVersion(emailTemplateId, 1)
+	if err != nil {
+		t.Fatalf("Error fetching template version: %v", err)
+	}
+
+	assertEqual(t, "go-sdk-test-email", resp.Name)
+	assertEqual(t, "email", resp.Type)
+	assertEqual(t, 1, resp.Version)
+	assertEqual(t, "Hi ((name)),\r\n\r\nMy favourite colour is ((colour)).", resp.Body)
+}
+
+func TestGenerateTemplatePreview(t *testing.T) {
+	client, _, emailTemplateId := setup(t)
+	resp, err := client.GenerateTemplatePreview(emailTemplateId, notify.Personalisation{
+		{"name", "KD"},
+		{"colour", "yellow"},
+	})
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
+	}
+
+	assertEqual(t, "Hi KD,\n\nMy favourite colour is yellow.\n", resp.Body)
+	assertEqual(t, "email", resp.Type)
+}
+
+func assertEqual(t *testing.T, expected, actual interface{}) {
```

Comment:
@andrewoh531 don't really even need to use this helper method, but if you're going to, more idiomatic is to reverse the argument order, use got and want and error message as follows:

```go
func assertEqual(t *testing.T, got, want interface{}) {
	if got != want {
		t.Errorf("got %v, want %v", got, want)
	}
}
```

The reason I'd not use it is because this for example:

```go
assertEqual(t, "email", resp.Type)
```

Would usually just be:

```go
want := "email" // And if this was a test from a table test, it would come off the table test, e.g. `tt.want` and so the local variable wouldn't be required.
if got := resp.Type; got != want {
	t.Errorf("Type = %v, want %v", got, want)
}
```

And now for a failed test, rather than reading generically, `got sms, want email` (or `Expected email but received sms`), it would read `Type = sms, want email` and so you now know what actually failed.

So in summary, after writing that, it's probably better to do this all inline and don't use the helper. Also the helper is really only a helper for a few types—it won't work if you try and pass it maps and slices, etc. In those cases, `reflect.DeepEqual()` is idiomatically used. With that in mind, `assertEqual` is a potential place for causing weird bugs in the tests.